zcash_client_sqlite/src/wallet/sapling.rs

```diff
@@ -1325,6 +1326,161 @@ pub(crate) mod tests {
+                recipient_address: RecipientAddress::Shielded(addr2),
+                amount: amount_sent,
+                memo: None,
+                label: None,
+                message: None,
+                other_params: vec![],
+            },
+            // payment back to the originating wallet, simulating legacy change
+            Payment {
+                recipient_address: RecipientAddress::Shielded(addr),
+                amount: Amount::from_u64(30000).unwrap(),
+                memo: None,
+                label: None,
+                message: None,
+                other_params: vec![],
+            },
+        ])
+        .unwrap();
+
+        let fee_rule = FixedFeeRule::standard();
+        let input_selector = GreedyInputSelector::new(
+            fixed::SingleOutputChangeStrategy::new(fee_rule),
+            DustOutputPolicy::default(),
+        );
+
+        let txid = spend(
+            &mut db_data,
+            &tests::network(),
+            test_prover(),
+            &input_selector,
+            &usk,
+            req,
+            OvkPolicy::Sender,
+            NonZeroU32::new(1).unwrap(),
+        )
+        .unwrap();
+        let tx = db_data.get_transaction(txid).unwrap();
+
+        let cb2 = fake_compact_block_from_tx(
+            sapling_activation_height() + 1,
+            BlockHash(cb.hash.try_into().unwrap()),
+            &tx,
+            cb.chain_metadata.unwrap().sapling_commitment_tree_size,
+        );
+        insert_into_cache(&db_cache, &cb2);
+        scan_cached_blocks(
+            &tests::network(),
+            &db_cache,
+            &mut db_data,
+            sapling_activation_height() + 1,
+            1,
+        )
+        .unwrap();
+
+        let amount_left = (value - (amount_sent + fee_rule.fixed_fee()).unwrap()).unwrap();
+        assert_eq!(
+            get_balance(&db_data.conn, AccountId::from(1)).unwrap(),
+            amount_sent
+        );
+        assert_eq!(
+            get_balance(&db_data.conn, AccountId::from(0)).unwrap(),
+            amount_left
+        );
+
+        let restore_file = NamedTempFile::new().unwrap();
+        let mut db_restore = WalletDb::for_path(restore_file.path(), tests::network()).unwrap();
+        init_wallet_db(&mut db_restore, None).unwrap();
+
+        assert_matches!(
+            scan_cached_blocks(
+                &tests::network(),
+                &db_cache, // same block cache
+                &mut db_restore,
+                sapling_activation_height(),
+                2,
+            ),
+            Ok(_)
+        );
+        assert_matches!(
+            decrypt_and_store_transaction(&tests::network(), &mut db_restore, &tx),
+            Ok(_)
+        );
+
+        assert_eq!(
+            get_balance(&db_restore.conn, AccountId::from(1)).unwrap(),
+            amount_sent
+        );
+        assert_eq!(
+            get_balance(&db_restore.conn, AccountId::from(0)).unwrap(),
+            amount_left
+        );
```

Comment:
These checks currently fail. I believe that's detecting the bug, but we'll have to fix the bug to be sure.

Reply:
You've initialized `db_restore`, but not added the account to the wallet, so this won't detect anything.